Apps/Src/BlinkLights.c

```diff
@@ -0,0 +1,93 @@
+#include "Tasks.h"
+#include "Lights.h"
+
+void Task_BlinkLight(void* p_arg){
+    //get necessary state pointers
+    car_state_t *car_state = (car_state_t*) p_arg;
+    blinker_states_t *blinkerStates = &(car_state -> BlinkerStates);
+    light_switches_t *storedLightSwitches = &(car_state -> SwitchStates.lightSwitches);
+    CPU_TS ts;
+    OS_ERR err;
+
+    while(1){
+        //wait for blinkLight to get signaled
+        OSSemPend( 
```

Comment:
If the blinkLights Sem4 isn't signaled, we should never go into this routine to blink the lights at all. That's why I've put it within the loop runtime. Do you think there's a better way to implement that?

Reply:
What is this semaphore used for?

Reply:
@ClarkPoon This semaphore gets signaled from the UpdateLights thread for which the source can be found [here](https://github.com/lhr-solar/Controls/blob/RTOS_UpdateLights/Apps/Src/UpdateLights.c). Essentially, if the Hazard switch or either of the indicator switches is toggled, that thread then signals this semaphore. This thread then controls the blinking on/off until it is no longer necessary.

Reply:
Responding to your TODO comment:

Personally, I would not use the semaphore here since the rate you are flashing the lights probably doesn't need to synchronize with the UpdateLights thread.

I am guessing you are using the semaphore because you are worried about `LightsSet()` wasting time by setting the lights when they don't need to be set. I think the semaphore works for this as long as you keep setting it at the same rate in your other task. One potentially cleaner solution to this would be:
```C
// have some array to keep track of the states of all of the lights
// this array would need to be initialized on startup
// NUM_LIGHTS can just be an extra at the end of your light_t enum
static State lights[NUM_LIGHTS];   

void Lights_Set(light_t light, State state) {
    if (state == lights[light]) {
        return;
    else {
        lights[light] = state;
    }

    // the rest of your lights code
    . . .
}
```